src/python/pants/reporting/json_reporter.py

```diff
@@ -0,0 +1,145 @@
+
+    # Results of the state of the build
+    self._results = {}
+    # Workunit stacks partitioned by thread
+    self._stack_per_thread = defaultdict(list)
+
+  def report_path(self):
+    """The path to the main report file."""
+
+    return os.path.join(self._json_dir, 'build.json')
+
+  def open(self):
+    """Implementation of Reporter callback."""
+
+    safe_mkdir(os.path.dirname(self._json_dir))
+    self._report_file = open(self.report_path(), 'w')
+
+  def close(self):
+    """Implementation of Reporter callback."""
+
+    if os.path.exists(self._json_dir):
+      self._report_file.write(json.dumps(
+        {
+          'workunits': self._results,
+          'artifact_cache_stats': self.run_tracker.artifact_cache_stats.get_all(),
+        }
+      ))
+      self._report_file.flush()
+    self._report_file.close()
+    for files in self._output_files.values():
+      for f in files.values():
+        f.close()
+
+  def start_workunit(self, workunit):
+    """Implementation of Reporter callback."""
+
+    workunit_data = {
+      'name': workunit.name,
+      'id': str(workunit.id),
+      'parent_name': workunit.parent.name if workunit.parent else '',
+      'parent_id': str(workunit.parent.id) if workunit.parent else '',
+      'labels': list(workunit.labels),
+      'cmd': workunit.cmd or '',
+      'start_time': workunit.start_time,
+      'outputs': defaultdict(str),
+      'children': [],
+      'log_entries': [],
+    }
+
+    root_id = str(workunit.root().id)
+
+    if not root_id in self._stack_per_thread:
+      self._stack_per_thread[root_id].append(workunit_data)
+      self._results[root_id] = workunit_data
+    else:
+      self._stack_per_thread[root_id][-1]['children'].append(workunit_data)
+      self._stack_per_thread[root_id].append(workunit_data)
+
+  def end_workunit(self, workunit):
+    """Implementation of Reporter callback."""
+
+    additional_data = {
+      'outcome': workunit.outcome_string(workunit.outcome()),
+      'end_time': workunit.end_time,
+      'unaccounted_time': workunit.unaccounted_time(),
+    }
+
+    root_id = str(workunit.root().id)
+
+    self._stack_per_thread[root_id][-1].update(additional_data)
+    self._stack_per_thread[root_id].pop()
+
+    for f in self._output_files[workunit.id].values():
+      f.close()
+
+  def handle_output(self, workunit, label, stream):
+    """Implementation of Reporter callback."""
+
+    self._stack_per_thread[str(workunit.root().id)][-1]['outputs'][label] += stream
+
+  def do_handle_log(self, workunit, level, *msg_elements):
+    """Implementation of Reporter callback."""
+
+    entry_info = {
+      'level': self._log_level_str[level],
+      'messages': self._render_messages(*msg_elements),
+    }
+
+    current_stack = self._stack_per_thread[str(workunit.root().id)]
+    if current_stack:
+      current_stack[-1]['log_entries'].append(entry_info)
+    else:
+      self._results[str(workunit.root().id)]['log_entries'].append(entry_info)
```

Comment:
It seems like this should probably apply to `workunit.id`, rather than `workunit.root().id`. Also, would you expect  `self._results(workunit.id) is current_stack[-1]` to be true?

Reply:
>It seems like this should probably apply to workunit.id, rather than workunit.root().id.

True, here `workunit.id` would be the same as `workunit.root().id` - I'll memoize it as we do in `start_workunit` instead since it still needs to be a string so we can serialize to JSON.

>Also, would you expect self._results(workunit.id) is current_stack[-1] to be true?

In this case the stack is empty (and the root workunit isn't included in that stack) - so we're appending any further log output directly to the root workunit results instead.